Add knot-cloud-core stack (#27)
CESARBR/knot-cloud: stacks/knot-cloud-core/README.md

# knot-cloud-core

### Configure DNS

This stack uses [Traefik](https://traefik.io) as a reverse proxy for the services and requires that you configure your DNS server to point, at least, the **things**, **users** and **bt** subdomains to the machine where the stack is being deployed, so that it can route the requests to the appropriated service. It is possible to configure it to route by path or port, but instructions for that won't be provided here for brevity.

If you don't have a domain or can't configure the main DNS server, you can configure a test domain in your machine before proceeding. Either setup a local DNS server, e.g. [bind9](https://wiki.debian.org/Bind9), or alternatively update your hosts file to include the following addresses:

```
127.0.0.1   things
127.0.0.1   users
127.0.0.1   bt
```

On Windows, the hosts file is usually located under `c:\Windows\System32\Drivers\etc\hosts`. On Unix systems, it is commonly found at `/etc/hosts`. Regardless of you operating system, administrator or super user privileges will be required.

### Deploy: stage 1

> Stage 1 contains the core services. The next sections provide the instructions to configure and deploy them. Whenever a configuration file is mentioned, it refers to a file found at `stacks/prod/env.d`, for production mode, or `<path>/stack/dev/env.d`, for development mode.

Reviewer: The path should be updated & we don't have neither `prod` for `knot-cloud-core` nor a `dev` being generated in `knot-cloud init knot-cloud-core`. :grimacing: 

```suggestion
Stage 1 contains the core services. The next sections provide the instructions to configure and deploy them. Whenever a configuration file is mentioned, it refers to a file found at `stacks/knot-cloud-core/dev/env.d` or `<path>/stack/env.d`.
```

Also related to this, why isn't the default path (for `knot-cloud init <stack>`) `knot-cloud/stacks/knot-cloud{-core}/dev/`? Also, what is the purpose of the pre-existing `/dev` dir?

It gets quite confusing to have `stack` and `stacks` on the same tree level, also having pre-existing env files for prod and dev while having a cli to generate even more env files :eyes:

Author: About your questions:

1. I don't really know. At least for me, the default should be the developer's `home` directory. This would avoid the `knot-cloud` directory becoming a mess.
2. The pre-existing `/dev` dir exists to avoid to emphasize that stack isn't suitable for development.

I agree with you that it gets confusing but creating inside the `dev` directory wouldn't improve that. I think the developer shouldn't use this directory as their workspace (only if really want).

#### Deploy

Deploy the stage 1 services:

```bash
docker stack deploy -c stage-1.yml knot-cloud-core
```

#### Verify

Wait until all the services are started. You can check it by running:

```bash
docker stack services knot-cloud-core
```

> And verifying that every service has one replica.

Reviewer: Personal opinion... this seems better to read, as you know everything you are looking for before running the command (avoids having alt+tabbing) :eyes: 

```suggestion
Check if all services are running and have exactly one replica:

```bash
docker stack services knot-cloud-core
```
>**Obs**: there should be a "```" on the last line of this suggestion (it's bugged).

Author: Done.

> And verifying that every service has one replica.

Reviewer: Just as we have for `knot-cloud` stack, we should add the `healthcheck` command for `babeltower` :point_down: 

```suggestion
And verifying that every service has one replica.

Make sure everything is running smoothly: 

```bash
$ curl https://bt.<your domain>/healthcheck
{ status: online}

**Note:** If the HTTPS certificates are not configured locally, traefik has a default certificate that is used in such cases. To use traefik's default certificate, it is necessary to add `-k` parameter to the `curl` command otherwise the request will fail.
```
>**Obs**: I'm not quite sure if the `<your_domain>` applies here. Usually, I only do `curl -k https://bt/healthcheck`. Also, there should be a "```" under the `curl` command output -- if I add it, the suggestion breaks.

Author: I can't agree with that because a lot of services will be added here and all of them have the same importance. The main difference is that the `knot-cloud` provides a `healthcheck` endpoint for the intermediate `bootstrapping` stage, the developer/devOps guy need to know this information to follow the next instructions.

Reviewer: I see your point and it makes sense :confetti_ball:
